examples/timeseries/timeseries_traffic_forecasting.py

```diff
@@ -0,0 +1,655 @@
+The function `create_tf_dataset` below takes as input a `numpy.ndarray` and returns a
+`tf.data.Dataset`. In this function `input_sequence_length=T` and `forecast_horizon=h`.
+The argument `multi_horizon` needs more explanation. Assume `forecast_horizon=3`.
+If `multi_horizon=True` then the model will make a forecast for time steps
+`t+T+1, t+T+2, t+T+3`. So the target will have shape `(T,3)`. But if
+`multi_horizon=False`, the model will make a forecast only for time step `t+T+3` and
+so the target will have shape `(T, 1)`.
+
+You may notice that the input tensor in each batch has shape
+`[batch_size, input_sequence_length, n_routes, 1]`. The last dimension is added to
+make the model more general: at each time step, the input features for each raod may
+contain multiple timeseries. For example, one might want to use temperature timeseries
+in addition to historical values of the speed as input features. In this notebook,
+however,
+the last dimension of the input is always 1.
+
+We use the last 12 values of the speed in each road to forecast the speed for 3 time
+steps ahead:
+"""
+
+from tensorflow.keras.preprocessing import timeseries_dataset_from_array
+
+batch_size = 64
+input_sequence_length = 12
+forecast_horizon = 3
+multi_horizon = False
+
+
+def create_tf_dataset(
+    data_array: np.ndarray,
+    input_sequence_length: int,
+    forecast_horizon: int,
+    batch_size: int = 128,
+    shuffle=True,
+    multi_horizon=True,
+):
+    """Creates tensorflow dataset from numpy array.
+
+This function creates a dataset where each element is a tuple (inputs, targets). `inputs`
+is a Tensor
+    of shape [batch_size, input_sequence_length, n_routes, 1] containing
+    the `input_sequence_length` past values of the timeseries for each node.
+`targets` is a Tensor of shape  [batch_size, forecast_horizon, n_routes] containing the
+`forecast_horizon`
+    future values of the timeseries for each node.
+
+    :param data_array: np.ndarray with shape [n_time_steps, n_routes]
+    :param input_sequence_length:
+    :param forecast_horizon:
+    :param batch_size:
+    :param shuffle:
+    :param multi_horizon:
+    :return:
+    """
+
+    inputs = timeseries_dataset_from_array(
+        np.expand_dims(data_array[:-forecast_horizon], axis=-1),
+        None,
+        sequence_length=input_sequence_length,
+        shuffle=False,
+        batch_size=batch_size,
+    )
+
+    target_offset = (
+        input_sequence_length
+        if multi_horizon
+        else input_sequence_length + forecast_horizon - 1
+    )
+    target_seq_length = forecast_horizon if multi_horizon else 1
+    targets = timeseries_dataset_from_array(
+        data_array[target_offset:],
+        None,
+        sequence_length=target_seq_length,
+        shuffle=False,
+        batch_size=batch_size,
+    )
+
+    dataset = tf.data.Dataset.zip((inputs, targets))
+    if shuffle:
+        dataset = dataset.shuffle(2048)
```

Comment:
Note that the number here is a number of batches since the data is already batched. 2048 seems excessive.